File: go/harness.go
package main

import (
	"fmt"
	"os"
	"runtime"
	"time"

	hdr "github.com/HdrHistogram/hdrhistogram-go"
)

// ---------------------------------------------------------------- harness

type report struct {
	fills    uint64
	volume   uint64
	resting  int
	checksum uint64
	hist     *hdr.Histogram
}

type runResult struct {
	rep     report
	wall    float64
	gcCount uint32
	gcPause time.Duration
	mallocs uint64
}

func run(totalOps, warmupOps, rate uint64, quiet bool) runResult {
	epoch := time.Now()
	rb := newRing()
	done := make(chan report, 1)

	tuned := os.Getenv("ENGINE") == "tuned"
	// GOGC is honored natively by the runtime; GOGC=off disables collection
	// entirely, which the tuned engine makes safe: with ~10k tiny allocations
	// per run in steady state the heap never grows. The only cycles a default
	// run sees are the pacer reacting to the book's construction. Printed so
	// every result line is self-documenting about the collector it ran under.
	gogc := os.Getenv("GOGC")
	if gogc == "" {
		gogc = "default"
	}
	if !quiet {
		fmt.Printf("  engine=%s gogc=%s\n",
			map[bool]string{true: "tuned", false: "idiomatic"}[tuned], gogc)
	}
	go func() {
		runtime.LockOSThread()
		var b engine
		if tuned {
			b = newTunedBook(totalOps)
		} else {
			b = newBook()
		}
		hist := hdr.New(1, 60_000_000_000, 3)
		seq := uint64(0)
		for {
			for rb.prod.Load() == seq {
				// busy-spin
			}
			seq++
			e := &rb.buf[seq&rb.mask]
			switch e.kind {
			case kindLimit:
				b.limit(e.side, e.price, e.qty, e.id)
			case kindCancel:
				b.cancel(e.id)
			default:
				f, v, r, c := b.stats()
				done <- report{f, v, r, c, hist}
				return
			}
			if e.ts != 0 {
				lat := uint64(time.Since(epoch).Nanoseconds()) - e.ts
				_ = hist.RecordValue(max(int64(lat), 1))
			}
			rb.cons.Store(seq)
		}
	}()

	runtime.LockOSThread()
	wl := newGen(42)
	var interval uint64
	if rate > 0 {
		interval = 1_000_000_000 / rate
	}
	nextDue := uint64(time.Since(epoch).Nanoseconds())

	var ms0, ms1 runtime.MemStats
	runtime.ReadMemStats(&ms0)
	t0 := time.Now()
	for i := uint64(0); i < totalOps; i++ {
		wl.nextOp()
		var ts uint64
		if interval > 0 {
			nextDue += interval
			for uint64(time.Since(epoch).Nanoseconds()) < nextDue {
				// busy-spin pacing
			}
			if i >= warmupOps {
				ts = nextDue
			}
		}
		e := rb.claim()
		e.kind, e.side = wl.kind, wl.side
		e.price, e.qty, e.id = wl.price, wl.qty, wl.id
		e.ts = ts
		rb.publish()
	}
	e := rb.claim()
	e.kind, e.ts = kindReport, 0
	rb.publish()
	rep := <-done
	wall := time.Since(t0).Seconds()
	runtime.ReadMemStats(&ms1)

	return runResult{
		rep:     rep,
		wall:    wall,
		gcCount: ms1.NumGC - ms0.NumGC,
		gcPause: time.Duration(ms1.PauseTotalNs - ms0.PauseTotalNs),
		mallocs: ms1.Mallocs - ms0.Mallocs,
	}
}

func printResult(r runResult, totalOps uint64, latency bool) {
	fmt.Printf("  gc: %d cycles, %.1f ms total stop-the-world pause; heap allocations: %d\n",
		r.gcCount, float64(r.gcPause.Nanoseconds())/1e6, r.mallocs)
	fmt.Printf("  ops=%d wall=%.3fs throughput=%.0f ops/s\n",
		totalOps, r.wall, float64(totalOps)/r.wall)
	fmt.Printf("  fills=%d volume=%d resting=%d checksum=%016x\n",
		r.rep.fills, r.rep.volume, r.rep.resting, r.rep.checksum)
	if latency {
		h := r.rep.hist
		us := func(q float64) float64 { return float64(h.ValueAtQuantile(q)) / 1000.0 }
		fmt.Printf("  latency µs: p50=%.3f p90=%.3f p99=%.3f p99.9=%.1f p99.99=%.1f max=%.1f (n=%d)\n",
			us(50), us(90), us(99), us(99.9), us(99.99),
			float64(h.Max())/1000.0, h.TotalCount())
	}
}

File: rust/src/harness.rs
//! Engine dispatch, the ring-buffer harness, and reporting.

use std::sync::atomic::Ordering;
use std::sync::{Arc, Mutex};
use std::time::Instant;

use disruptor::{build_single_producer, BusySpin, Producer, ProcessorSettings};
use hdrhistogram::Histogram;

use crate::alloc::ALLOCS;
use crate::book::Book;
use crate::tuned::TunedBook;
use crate::workload::{Gen, KIND_CANCEL, KIND_LIMIT, KIND_REPORT};

// engine dispatch: idiomatic (default) or tuned (ENGINE=tuned)

enum AnyBook {
    Idio(Book),
    Tuned(Box<TunedBook>),
}

impl AnyBook {
    fn limit(&mut self, side: u8, price: i64, qty: i64, id: u64) {
        match self {
            AnyBook::Idio(b) => b.limit(side, price, qty, id),
            AnyBook::Tuned(b) => b.limit(side, price, qty, id),
        }
    }
    fn cancel(&mut self, id: u64) {
        match self {
            AnyBook::Idio(b) => b.cancel(id),
            AnyBook::Tuned(b) => b.cancel(id),
        }
    }
    fn stats(&self) -> (u64, u64, usize, u64) {
        match self {
            AnyBook::Idio(b) => (b.fills, b.volume, b.orders.len(), b.checksum()),
            AnyBook::Tuned(b) => (b.fills, b.volume, b.orders.len(), b.checksum()),
        }
    }
}

// ---------------------------------------------------------------- harness

struct Event {
    kind: u8,
    side: u8,
    price: i64,
    qty: i64,
    id: u64,
    ts: u64, // scheduled-publish nanos since run epoch; 0 = don't measure
}

pub struct Report {
    fills: u64,
    volume: u64,
    resting: usize,
    checksum: u64,
    hist: Histogram<u64>,
}

struct Core {
    book: AnyBook,
    hist: Histogram<u64>,
    epoch: Instant,
    report: Arc<Mutex<Option<Report>>>,
}

impl Core {
    fn on_event(&mut self, e: &Event) {
        match e.kind {
            KIND_LIMIT => self.book.limit(e.side, e.price, e.qty, e.id),
            KIND_CANCEL => self.book.cancel(e.id),
            _ => {
                let (fills, volume, resting, checksum) = self.book.stats();
                *self.report.lock().unwrap() =
                    Some(Report { fills, volume, resting, checksum, hist: self.hist.clone() });
                return;
            }
        }
        if e.ts != 0 {
            let now = self.epoch.elapsed().as_nanos() as u64;
            let _ = self.hist.record(now.saturating_sub(e.ts).max(1));
        }
    }
}

fn env_core(name: &str) -> Option<usize> {
    std::env::var(name).ok().and_then(|v| v.parse().ok())
}

fn engine_from_env() -> (AnyBook, &'static str) {
    if std::env::var("ENGINE").as_deref() == Ok("tuned") {
        (AnyBook::Tuned(Box::new(TunedBook::new())), "tuned")
    } else {
        (AnyBook::Idio(Book::default()), "idiomatic")
    }
}

// Power of two, as the disruptor crate requires: the ring indexes slots with
// `seq & (N-1)` — a single AND computing seq mod N, not a division.
const RING_SIZE: usize = 65_536;

pub fn run(total_ops: u64, warmup_ops: u64, rate: Option<u64>, quiet: bool) -> (Report, f64) {
    let epoch = Instant::now();
    let report: Arc<Mutex<Option<Report>>> = Arc::new(Mutex::new(None));
    let (book, engine) = engine_from_env();
    let pin_prod = env_core("PIN_PROD");
    let pin_cons = env_core("PIN_CONS");
    if !quiet {
        println!(
            "  engine={engine} pin_prod={} pin_cons={}",
            pin_prod.map_or("-".into(), |c| c.to_string()),
            pin_cons.map_or("-".into(), |c| c.to_string())
        );
    }
    let mut core = Core {
        book,
        hist: Histogram::new_with_bounds(1, 60_000_000_000, 3).unwrap(),
        epoch,
        report: Arc::clone(&report),
    };

    let factory = || Event { kind: 0, side: 0, price: 0, qty: 0, id: 0, ts: 0 };
    let handler = move |e: &Event, _seq: i64, _eob: bool| core.on_event(e);
    let builder = build_single_producer(RING_SIZE, factory, BusySpin);
    let mut producer = match pin_cons {
        Some(c) => builder.pin_at_core(c).handle_events_with(handler).build(),
        None => builder.handle_events_with(handler).build(),
    };
    // Pin the producer only after the consumer thread exists: on Linux, child
    // threads inherit the parent's affinity mask, so pinning first would leave
    // the consumer no core to be pinned to.
    if let Some(c) = pin_prod {
        core_affinity::set_for_current(core_affinity::CoreId { id: c });
    }

    let mut wl = Gen::new(42);
    let interval = rate.map(|r| 1_000_000_000 / r);
    let mut next_due: u64 = epoch.elapsed().as_nanos() as u64;

    let allocs0 = ALLOCS.load(Ordering::Relaxed);
    let t0 = Instant::now();
    for i in 0..total_ops {
        let op = wl.next_op();
        let ts = match interval {
            Some(iv) => {
                next_due += iv;
                while (epoch.elapsed().as_nanos() as u64) < next_due {
                    std::hint::spin_loop();
                }
                if i >= warmup_ops { next_due } else { 0 }
            }
            None => 0,
        };
        producer.publish(|e| {
            e.kind = op.kind;
            e.side = op.side;
            e.price = op.price;
            e.qty = op.qty;
            e.id = op.id;
            e.ts = ts;
        });
    }
    producer.publish(|e| {
        e.kind = KIND_REPORT;
        e.ts = 0;
    });
    drop(producer); // joins the consumer thread
    let wall = t0.elapsed().as_secs_f64();
    let allocs = ALLOCS.load(Ordering::Relaxed) - allocs0;

    let rep = report.lock().unwrap().take().expect("consumer did not report");
    if !quiet {
        println!("  heap allocations during run: {allocs}");
    }
    (rep, wall)
}

pub fn print_report(rep: &Report, wall: f64, total_ops: u64, latency: bool) {
    println!(
        "  ops={} wall={:.3}s throughput={:.0} ops/s",
        total_ops,
        wall,
        total_ops as f64 / wall
    );
    println!(
        "  fills={} volume={} resting={} checksum={:016x}",
        rep.fills, rep.volume, rep.resting, rep.checksum
    );
    if latency {
        let h = &rep.hist;
        let us = |q: f64| h.value_at_quantile(q) as f64 / 1_000.0;
        println!(
            "  latency µs: p50={:.3} p90={:.3} p99={:.3} p99.9={:.1} p99.99={:.1} max={:.1} (n={})",
            us(0.50),
            us(0.90),
            us(0.99),
            us(0.999),
            us(0.9999),
            h.max() as f64 / 1_000.0,
            h.len()
        );
    }
}

